bin/undraft.ml

```diff
@@ -0,0 +1,222 @@
+open Bos_setup
+open Dune_release
+
+let get_pkg_dir pkg =
+  Pkg.build_dir pkg >>= fun bdir ->
+  Pkg.distrib_filename ~opam:true pkg >>= fun fname -> Ok Fpath.(bdir // fname)
+
+let pp_opam_repo fmt opam_repo =
+  let user, repo = opam_repo in
+  Format.fprintf fmt "%s/%s" user repo
+
+let update_opam_file ~dry_run ~url pkg =
+  get_pkg_dir pkg >>= fun dir ->
+  Pkg.opam pkg >>= fun opam_f ->
+  OS.Dir.create dir >>= fun _ ->
+  let dest_opam_file = Fpath.(dir / "opam") in
+  let url = OpamUrl.parse url in
+  Pkg.distrib_file ~dry_run pkg >>= fun distrib_file ->
+  let file = Fpath.to_string distrib_file in
+  let hash algo = OpamHash.compute ~kind:algo file in
+  let checksum = List.map hash [ `SHA256; `SHA512 ] in
+  let url = OpamFile.URL.create ~checksum url in
+  OS.File.read opam_f >>= fun opam ->
+  let opam_t = OpamFile.OPAM.read_from_string opam in
+  (match OpamVersion.to_string (OpamFile.OPAM.opam_version opam_t) with
+  | "2.0" ->
+      let file x = OpamFile.make (OpamFilename.of_string (Fpath.to_string x)) in
+      let opam_t = OpamFile.OPAM.with_url url opam_t in
+      if not dry_run then
+        OpamFile.OPAM.write_with_preserved_format ~format_from:(file opam_f)
+          (file dest_opam_file) opam_t;
+      Ok ()
+  | ("1.0" | "1.1" | "1.2") as v ->
+      App_log.status (fun l ->
+          l "Upgrading opam file %a from opam format %s to 2.0" Text.Pp.path
+            opam_f v);
+      let opam =
+        OpamFile.OPAM.with_url url opam_t |> OpamFile.OPAM.write_to_string
+      in
+      Sos.write_file ~dry_run dest_opam_file opam
+  | s -> Fmt.kstrf (fun x -> Error (`Msg x)) "invalid opam version: %s" s)
+  >>| fun () ->
+  App_log.success (fun m ->
+      m "Wrote opam package description %a" Text.Pp.path dest_opam_file)
+
+let undraft ?opam ?distrib_uri ?distrib_file ?opam_repo ?user ?token ?local_repo
+    ?remote_repo ?build_dir ?pkg_names ~dry_run ~yes:_ () =
+  let pkg = Pkg.v ?opam ?distrib_file ?build_dir ~dry_run:false () in
+  Pkg.name pkg >>= fun pkg_name ->
+  Pkg.build_dir pkg >>= fun build_dir ->
+  Pkg.version pkg >>= fun version ->
+  Pkg.tag pkg >>= fun tag ->
+  let pkg_names = match pkg_names with Some x -> x | None -> [] in
+  let pkg_names = pkg_name :: pkg_names in
+  let opam_repo =
+    match opam_repo with None -> ("ocaml", "opam-repository") | Some r -> r
+  in
+  Config.v ~user ~local_repo ~remote_repo [ pkg ] >>= fun config ->
+  (match local_repo with
+  | Some r -> Ok Fpath.(v r)
+  | None -> (
+      match config.local with
+      | Some r -> Ok r
+      | None -> R.error_msg "Unknown local repository."))
+  >>= fun local_repo ->
+  (match remote_repo with
+  | Some r -> Ok r
```

Comment:
I tried a few things but the mixing of the result and the option types make it not easier to read. I think it can be improved in a following PR, I would like to do something to what we have in ocamlformat, that is: a Config.t type that is built by overriding the default config with environment, .ocamlformat files and options passed by the user, for dnue-release it can be: current config.t read from the yaml file + command line options, it should be computed once at the beginning of each command and then just passed down each function wihtout having to worry if we have taken each option into account.
What do you think?

(The other way would be to dramatically reduce the command line options of dune-release)